Userland/Libraries/LibCompress/Lzma.cpp

```diff
@@ -0,0 +1,604 @@
+
+    m_range_decoder_code += m_range_decoder_range & t;
+
+    if (m_range_decoder_code == m_range_decoder_range)
+        return Error::from_string_literal("Reached an invalid state while decoding LZMA stream");
+
+    TRY(normalize_range_decoder());
+
+    // The reference implementation adds "t + 1" to the accumulated return value, so that should presumably be either 0 or 1.
+    VERIFY(t + 1 == 0 || t + 1 == 1);
+    return t + 1;
+}
+
+ErrorOr<u8> LzmaDecompressor::decode_bit_with_probability(Probability& probability)
+{
+    // "The LZMA decoder provides the pointer to CProb variable that contains
+    //  information about estimated probability for symbol 0 and the Range Decoder
+    //  updates that CProb variable after decoding."
+
+    // The significance of the shift width is not explained and appears to be a magic constant.
+    constexpr size_t probability_shift_width = 5;
+
+    u32 bound = (m_range_decoder_range >> probability_bit_count) * probability;
+
+    if (m_range_decoder_code < bound) {
+        probability += ((1 << probability_bit_count) - probability) >> probability_shift_width;
+        m_range_decoder_range = bound;
+        TRY(normalize_range_decoder());
+        return 0;
+    } else {
+        probability -= probability >> probability_shift_width;
+        m_range_decoder_code -= bound;
+        m_range_decoder_range -= bound;
+        TRY(normalize_range_decoder());
+        return 1;
+    }
+}
+
+ErrorOr<u16> LzmaDecompressor::decode_symbol_using_bit_tree(size_t bit_count, Span<Probability> probability_tree)
+{
+    VERIFY(bit_count <= sizeof(u16) * 8);
+    VERIFY(probability_tree.size() >= 1ul << bit_count);
+
+    // This has been modified from the reference implementation to unlink the result and the tree index,
+    // which should allow for better readability.
+
+    u16 result = 0;
+    size_t tree_index = 1;
+
+    for (size_t i = 0; i < bit_count; i++) {
+        u16 next_bit = TRY(decode_bit_with_probability(probability_tree[tree_index]));
+        result = (result << 1) | next_bit;
+        tree_index = (tree_index << 1) | next_bit;
+    }
+
+    return result;
+}
+
+ErrorOr<u16> LzmaDecompressor::decode_symbol_using_reverse_bit_tree(size_t bit_count, Span<Probability> probability_tree)
+{
+    VERIFY(bit_count <= sizeof(u16) * 8);
+    VERIFY(probability_tree.size() >= 1ul << bit_count);
+
+    u16 result = 0;
+    size_t tree_index = 1;
+
+    for (size_t i = 0; i < bit_count; i++) {
+        u16 next_bit = TRY(decode_bit_with_probability(probability_tree[tree_index]));
+        result |= next_bit << i;
+        tree_index = (tree_index << 1) | next_bit;
+    }
+
+    return result;
+}
+
+ErrorOr<void> LzmaDecompressor::decode_literal_to_output_buffer()
+{
+    u8 previous_byte = 0;
+    if (m_total_decoded_bytes > 0) {
+        auto read_bytes = MUST(m_output_buffer.read_with_seekback({ &previous_byte, sizeof(previous_byte) }, 1));
+        VERIFY(read_bytes.size() == sizeof(previous_byte));
```

Comment:
Not sure I understand this verify, if `read_with_seekback` didn't return an error `read_bytes` is a guarenteed to return 1 byte right? Otherwise it's a bug in the `read_with_seekback` implementation? If so, this seems a bit excessive, we should either assert this inside the implementation itself, or not at all.

Reply:
(same for everywhere else below)